handler/src/main/java/io/netty/handler/ssl/SniHandler.java

```diff
@@ -313,31 +333,58 @@ public void operationComplete(Future<SslContext> future) throws Exception {
     }
 
     /**
-     * The default implementation will simply call {@link AsyncMapping#map(Object, Promise)} but
-     * users can override this method to implement custom behavior.
+     * The default implementation will simply call {@link SslContextMapping#map(String, Promise)} but users can
+     * override this method to implement custom behavior.
      *
-     * @see AsyncMapping#map(Object, Promise)
+     * NOTE: This method returns a {@code Future<?>} to accommodate the old and new way of doing things without
+     * braking the API. Users who override this method are advised to change their return value from
+     * {@code Future<SslContext>} to {@code Future<Supplier<? extends SslContext>>}.
+     *
+     * @see SslContextMapping#map(String, Promise)
+     * @deprecated the return value will at some point change from {@code Future<SslContext>} to
```

Comment:
Originally in https://github.com/netty/netty/pull/6603#discussion_r110068929 but the conversation was collapsed

me
> can you add this new method now, reference it from the deprecated statement, and only use the new method so we have stronger type guarantees with the generics and less casting?

@rkapsi 
> Not without breaking the API but it's possible that we're the only ones overriding it in our extended SniHandler. 🤔

I don't think we have to break the API. What about just adding a new method with a different name?

```java
    /**
     * The default implementation will simply call {@link SslContextMapping#map(Object, Promise)} but users can
     * override this method to implement custom behavior.
     *
     * NOTE: This method returns a {@code Future<?>} to accommodate the old and new way of doing things without
     * braking the API. Users who override this method are advised to change their return value from
     * {@code Future<SslContext>} to {@code Future<Supplier<? extends SslContext>>}.
     *
     * @see SslContextMapping#map(Object, Promise)
     * @deprecated Use {@link #lookupSupplier(io.netty.channel.ChannelHandlerContext, String)}.
     */
    @Deprecated
    protected Future<SslContext> lookup(ChannelHandlerContext ctx, String hostname) throws Exception {
        final Promise<SslContext> promise = ctx.executor().newPromise();
        lookupSupplier(ctx, hostname).addListener(new GenericFutureListener<Future<Supplier<? extends SslContext>>>() {
            @Override
            public void operationComplete(Future<Supplier<? extends SslContext>> future) throws Exception {
                if (future.isSuccess()) {
                    // todo: do we care about null here, or if the supplier throws?
                    promise.setSuccess(future.getNow().get());
                } else if (future.isCancelled()) {
                    promise.cancel(false);
                } else {
                    promise.setFailure(future.cause());
                }
            }
        });
        return promise;
    }

    protected Future<Supplier<? extends SslContext>> lookupSupplier(ChannelHandlerContext ctx,
                                                                    String hostname) throws Exception {
        Promise<Supplier<? extends SslContext>> promise = ctx.executor().newPromise();
        return mapping.map(hostname, promise);
    }
```